### PR TITLE
delivers a will waits for authorization

### PR DESCRIPTION
**Pull request by nguyenthenguyen**
Fixed for crash at [comment](https://github.com/mcollina/aedes/pull/212/files#diff-50cfa59973c04321b5da0c6da0fdf4feL253)

### CODE REVIEW

File: lib/client.js [hunk 1]
```diff
@@ -251,14 +251,14 @@ Client.prototype.close = function (done) {
...
           if (!err) {
+            that.will = null // this function might be called twice
             that.broker.persistence.delWill({
               id: that.id,
               brokerId: that.broker.id
             }, nop)
           }
         }
       })
-      that.will = null // this function might be called twice
```
Comment (mhverbakel) [inline lib/client.js]: Removing this line here allows the function to be called twice. You should leave this line, and make an in-closure variable for `that.will`.

File: lib/client.js [hunk 2]
```diff
@@ -258,7 +258,7 @@ Client.prototype.close = function (done) {
...
+      that._finished = true // this function might be called twice
```
Comment (mcollina) [inline lib/client.js]: a better solution is to extract `that.will` into a variable, and keep that into the context of the closure.

File: test/will.js [hunk 3]
```diff
@@ -184,6 +184,28 @@ test('delivers a will with authorization', function (t) {
   s.conn.destroy()
 })
 
+test('delivers a will waits for authorization', function (t) {
+  let authorized = false
+  var opts = {}
+  // willConnect populates opts with a will
+  var s = willConnect(setup(aedes({ authorizePublish: (_1, _2, callback) => { authorized = true; setTimeout(() => { callback(null) }, 10) } })), opts)
```
Comment (mcollina) [inline test/will.js]: Can you shorten up the line and use a `setImmediate` instead?

File: lib/client.js (hunk 2 above)
Comment (nguyenthenguyen) [reply lib/client.js]: @mcollina Thanks you, sir!

**Pull request merged**